### DISCUSSION TITLE
Parsing edi file with interchange & transaction message => jaxb

### DISCUSSION THREAD

**Discussion by vadim-trotno** [xlate/staedi#249] [Q&A] [answered]
Hi ... first of all thanks for this powerful tool <3

I am currently trying to import an edi file ([AO06Q122.zip](https://github.com/xlate/staedi/files/7992877/AO06Q122.zip)) with the following structure:

- UNA
- UNB
  - UNH
    - CUSTOM_TRANSACTION Message
   - UNT
- UNZ

So in my case it seems like i need the interchange schema + a separate transaction schema.

My goal is to import this as a pojo with jaxb. 

Here is a code snippet:

        try (InputStream stream = new FileInputStream(filePath);
             EDIStreamReader reader = factory.createEDIStreamReader(stream, controlSchema)) {
            XMLStreamReader xmlStreamReader = factory.createXMLStreamReader(reader);
            while (xmlStreamReader.hasNext()) {
                if (xmlStreamReader.nextTag() == XMLStreamConstants.START_ELEMENT) {
                    JAXBContext context = JAXBContext.newInstance(Interchange.class);
                    Unmarshaller u = context.createUnmarshaller();
                    Interchange unmarshal = (Interchange) u.unmarshal(xmlStreamReader);
                    transactions.addAll(unmarshal.getTransaction());
                }
            }
        } catch (IOException | XMLStreamException e) {
            log.error("Exception while parsing file.", e);
            return new ArrayList<>();
        }
        
        return transactions.stream()
            .map(t -> mapTransactionToPayer())
            .collect(Collectors.toList());

My Problem is now that i dont know how to pass the transaction schema to the XMLStreamReader. 
I already tried it with the EDIReader approach... but then i am not able to utilize the XMLStreamReader.
Do you have an idea?

Here is a my code snippet with a combination of edi and xml reader:

        EDIInputFactory factory = EDIInputFactory.newFactory();
        try (InputStream stream = new FileInputStream(filePath);
             EDIStreamReader reader = factory.createEDIStreamReader(stream)) {
    
            ArrayList<Transaction> transactions = new ArrayList<>();
    
            while (reader.hasNext()) {
                switch (reader.next()) {
                    case START_TRANSACTION:
                        SchemaFactory schemaFactory = SchemaFactory.newFactory();
                        Schema txSchema = schemaFactory.createSchema(new FileInputStream("./payer_transaction_schema.xml"));
                        reader.setTransactionSchema(txSchema);

                        transactions.add(parseTransaction(factory, reader));
                        break;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        } 

    private Transaction parseTransaction(EDIInputFactory factory, EDIStreamReader reader) throws XMLStreamException, JAXBException 
    {
        try {
            XMLStreamReader xmlStreamReader = factory.createXMLStreamReader(reader);
            // exception occurs with or without this next() call
            xmlStreamReader.next();
            JAXBContext context = JAXBContext.newInstance(Transaction.class);
            Unmarshaller u = context.createUnmarshaller();
            return (Transaction) u.unmarshal(xmlStreamReader);
        } catch (Exception e) {
            return null;
        }
    }

Edit:
I currently workaround this problem by using the flat hierarchy of the XML.

**Comment by MikeEdgar** [comment, marked as answer]
@vadim-trotno You can accomplish this using an `EDIStreamFilter`. After you build the `EDIStreamReader`, you can wrap it with a filter that will set the schema at the relevant event. This will intercept the reader as it feeds EDI data into the XML reader that was also created from the factory.

```java
EDIStreamReader reader = factory.createEDIStreamReader(stream);
EDIStreamFilter filter = new MyCustomFilter();
EDIStreamReader filteredReader = factory.createFilteredReader(reader, filter);
```

```java
class MyCustomFilter implements EDIStreamFilter {
  public boolean accept(EDIStreamReader reader) {
    if (reader.getEventType() == EDIStreamEvent.START_TRANSACTION) {
      SchemaFactory schemaFactory = SchemaFactory.newFactory();
      Schema txSchema = schemaFactory.createSchema(new FileInputStream("./payer_transaction_schema.xml"));
      reader.setTransactionSchema(txSchema);
    }
    return true; // allowing all events through filter
  }
}
```

**Reply by vadim-trotno** [reply]
Thank you for you response/answer. I will use it that way!
